In [88]:
import pandas as pd

In [89]:
df = pd.read_csv('data/raw_master_dataset.csv')
print("Read raw master csv success")
display(df.head(5))
display(df.info())
display(df.describe())
display(df['year'].unique().tolist())
print('='*60)
print(f'Data time {df["year"].unique().min()} - {df["year"].unique().max()}')
print(f'Data shape {df.shape}')
print(f'Data size {df.size}')
print(f'Missing value {df[df.isnull().any(axis=1)].size} (~{(df[df.isnull().any(axis=1)].size / df.size):.2f}%)')
print(f'Total Country {df["country"].nunique()}')
print('='*60)


Read raw master csv success


,country_code,country,year,co2_per_capita,co2,ef_per_capita,BANK,ATM,DEP,LOAN,...,URB,FDI,GDP,NTR,PO,TO,IND,INF,PS,ED
0,ABW,Aruba,2010,25.030,2.506,NaN,NaN,NaN,83.440985,60.478900,...,64.161297,7.611672,25429.218097,0.003870,101838.0,135.464655,11.546062,-1.223407,1.256067,0.0014
1,ABW,Aruba,2011,24.706,2.495,NaN,20.776300,109.992178,74.539696,58.043001,...,63.999261,18.505780,26093.054208,0.003810,102591.0,154.437787,12.366448,4.005674,1.251139,0.0014
2,ABW,Aruba,2012,13.212,1.345,NaN,21.542954,117.289418,81.439475,60.680613,...,63.825209,-12.033180,25444.733665,0.003408,104110.0,156.645197,10.076063,0.184033,1.315909,0.0014
3,ABW,Aruba,2013,8.395,0.861,NaN,18.783973,120.921823,77.909008,61.768496,...,63.638291,9.974457,26680.141121,0.002452,105675.0,155.731905,10.749665,-1.995948,1.314319,0.0014
4,ABW,Aruba,2014,8.435,0.872,NaN,18.532890,114.672257,78.149919,62.624559,...,63.439253,7.465255,26773.822910,0.003235,106807.0,156.491053,10.799498,0.871011,1.210548,0.0014


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4926 entries, 0 to 4925
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   country_code    4386 non-null   object 
 1   country         3825 non-null   object 
 2   year            4926 non-null   int64  
 3   co2_per_capita  3480 non-null   float64
 4   co2             3720 non-null   float64
 5   ef_per_capita   1774 non-null   float64
 6   BANK            2702 non-null   float64
 7   ATM             2670 non-null   float64
 8   DEP             2618 non-null   float64
 9   LOAN            2622 non-null   float64
 10  FT_Mobile       3635 non-null   float64
 11  FT_Broadband    3540 non-null   float64
 12  FT_Internet     3360 non-null   float64
 13  LCE             3084 non-null   float64
 14  URB             4160 non-null   float64
 15  FDI             3687 non-null   float64
 16  GDP             3996 non-null   float64
 17  NTR             3007 non-null   f

None

,year,co2_per_capita,co2,ef_per_capita,BANK,ATM,DEP,LOAN,FT_Mobile,FT_Broadband,...,URB,FDI,GDP,NTR,PO,TO,IND,INF,PS,ED
count,4926.000000,3480.000000,3720.000000,1774.000000,2702.000000,2670.000000,2.618000e+03,2.622000e+03,3635.000000,3540.000000,...,4160.000000,3687.000000,3996.000000,3007.000000,4.160000e+03,3465.000000,3837.000000,4000.000000,3087.000000,3672.000000
mean,2017.423873,4.972782,932.062517,3.058113,17.628859,49.565352,8.603638e+09,5.056974e+09,104.072300,13.919901,...,60.036095,7.370909,15555.858394,6.042942,2.836733e+08,86.844179,26.066983,69.801259,-0.011522,161.450245
std,4.570907,5.843551,3676.396442,2.345742,19.412260,47.247595,1.839450e+11,1.065186e+11,39.486765,14.056336,...,22.622741,58.935809,22535.063160,9.047058,9.426432e+08,55.305069,11.537434,3584.761571,0.988212,508.361268
min,2010.000000,0.022000,0.000000,0.272142,0.225712,0.091334,3.747879e+00,4.818802e-01,1.211642,0.000000,...,10.984310,-1303.108267,253.308578,0.000000,9.492000e+03,0.000000,0.000000,-42.918604,-3.016512,0.000000
25%,2013.000000,0.977750,1.693750,1.262896,5.786877,12.688027,2.990277e+01,2.107453e+01,80.589390,1.145113,...,41.733696,1.222373,2108.270825,0.261915,1.502597e+06,52.022995,18.608987,1.550035,-0.650647,0.572950
50%,2017.000000,3.360000,13.419500,2.386854,13.251325,40.214982,4.716806e+01,4.092368e+01,106.328338,9.310077,...,60.369000,2.511559,6139.726300,2.201002,1.007072e+07,73.537288,25.039083,3.459072,0.031523,5.282750
75%,2021.000000,6.710750,150.618750,4.256564,22.788157,66.609790,7.475558e+01,6.246650e+01,125.610500,24.891969,...,78.555595,4.625017,18982.361768,7.885171,5.175450e+07,105.130362,31.846858,6.444630,0.842078,39.797300
max,2025.000000,49.152000,38598.578000,17.643144,259.251367,317.785383,5.107645e+12,2.619297e+12,574.152333,75.749699,...,100.000000,1709.827232,247170.103807,79.430949,8.215425e+09,679.232773,78.899528,225652.234994,1.759751,4333.608958


[2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018,
 2019,
 2020,
 2021,
 2022,
 2023,
 2024,
 2025]

Data time 2010 - 2025
Data shape (4926, 24)
Data size 118224
Missing value 95376 (~0.81%)
Total Country 254


# Cleaning data 
As you can see total countries is 254 so we need to narrow that number down to 145 countries net zero 

In [ ]:
# Danh sách 145 quốc gia Net-Zero (kèm chuẩn hóa tên quốc gia theo chuẩn ISO / World Bank / OWID)
net_zero_countries = [
    "Argentina", "Armenia", "Antigua and Barbuda", "Austria", "Belgium", "Burkina Faso", "Bangladesh", "Belize", "Barbados", "Bhutan", 
    "Central African Republic", "Canada", "Switzerland", "Chile", "China", "Colombia", "Comoros", "Cape Verde", "Costa Rica", "Cyprus", 
    "Czech Republic", "Czechia", "Germany", "Djibouti", "Dominica", "Denmark", "Dominican Republic", "Spain", "Estonia", "Ethiopia", "Finland", 
    "Fiji", "France", "Micronesia", "Micronesia (country)", "Gabon", "United Kingdom", "Gambia", "Greece", "Croatia", "Haiti", "Hungary", 
    "Ireland", "Iceland", "Italy", "Jamaica", "Japan", "Cambodia", "Kiribati", "Saint Kitts and Nevis", "South Korea", "Laos", 
    "Lebanon", "Liberia", "Lesotho", "Lithuania", "Luxembourg", "Latvia", "Monaco", "Mexico", "Marshall Islands", "Mali", 
    "Malta", "Myanmar", "Mozambique", "Mauritania", "Mauritius", "Malawi", "Namibia", "Niger", "Nepal", "Nauru", 
    "New Zealand", "Pakistan", "Peru", "Palau", "Papua New Guinea", "Portugal", "Rwanda", "Solomon Islands", "Sao Tome and Principe", "Suriname", 
    "Slovakia", "Slovenia", "Sweden", "Seychelles", "Togo", "East Timor", "Tonga", "Trinidad and Tobago", "Tuvalu", "Uganda", 
    "Uruguay", "Vanuatu", "Samoa", "Yemen", "South Africa", "Andorra", "United Arab Emirates", "Australia", "Bulgaria", "Bahrain", 
    "Brazil", "Brunei", "Georgia", "Ghana", "Honduras", "Indonesia", "India", "Israel", "Kazakhstan", "Kyrgyzstan", 
    "Kuwait", "Liechtenstein", "Sri Lanka", "Moldova", "Montenegro", "Mongolia", "Malaysia", "Nigeria", "Netherlands", "Oman", 
    "Panama", "Romania", "Russia", "Saudi Arabia", "Sudan", "Thailand", "Tunisia", "Turkey", "Tanzania", "Ukraine", 
    "Uzbekistan", "Venezuela", "Vietnam", "Singapore", "United States", "Poland", "Jordan", "Philippines", 
    "Kenya", "Ecuador", "Iraq", "Zambia", "Iran", "Morocco", "Norway"
]

# Chuẩn hóa tên nước Séc và Micronesia trong DataFrame về tên người dùng nhập
# Lọc danh sách quốc gia Net-Zero
filtered_df = df[df['country'].isin(net_zero_countries)].copy()
# Chuẩn hóa hiển thị tên
filtered_df['country'] = filtered_df['country'].replace({'Czechia': 'Czech Republic', 'Micronesia (country)': 'Micronesia'})

# Lưu ra file master_dataset.csv đã được làm sạch
filtered_df.to_csv('data/master_dataset.csv', index=False, encoding='utf-8')
df = filtered_df

unique_year = df['year'].unique()
missing_cells = df.isnull().sum().sum()
total_cells = df.size
percentage = (missing_cells / total_cells) * 100

print('Save master_dataset.csv clean countries assign NET ZERO')
print(f'Khung thời gian (Time range)      : {unique_year.min()} - {unique_year.max()}')
print(f'Tổng số ô dữ liệu (Total cells)   : {df.size} ô')
print(f'Kích thước tập dữ liệu (Shape)    : {df.shape} [{df.shape[0]} dòng x {df.shape[1]} cột]')
print(f'Tổng số quốc gia Net-Zero         : {df["country"].nunique()}')
print(f'Tổng số ô khuyết (Missing cells)  : {missing_cells} ô (~{percentage:.2f}%)')


Save master_dataset.csv clean countries assign NET ZERO
Khung thời gian (Time range)      : 2010 - 2024
Tổng số ô dữ liệu (Total cells)   : 52560 ô
Kích thước tập dữ liệu (Shape)    : (2190, 24) [2190 dòng x 24 cột]
Tổng số ô dữ liệu (Total cells)   : 52560 ô
Tổng số quốc gia Net-Zero         : 145
Tổng số ô khuyết (Missing cells)  : 3703 ô (~7.05%)


In [94]:
df['country'].unique().tolist()

['Andorra',
 'United Arab Emirates',
 'Argentina',
 'Armenia',
 'Antigua and Barbuda',
 'Australia',
 'Austria',
 'Belgium',
 'Burkina Faso',
 'Bangladesh',
 'Bulgaria',
 'Bahrain',
 'Belize',
 'Brazil',
 'Barbados',
 'Brunei',
 'Bhutan',
 'Central African Republic',
 'Canada',
 'Switzerland',
 'Chile',
 'China',
 'Colombia',
 'Comoros',
 'Cape Verde',
 'Costa Rica',
 'Cyprus',
 'Czech Republic',
 'Germany',
 'Djibouti',
 'Dominica',
 'Denmark',
 'Dominican Republic',
 'Ecuador',
 'Spain',
 'Estonia',
 'Ethiopia',
 'Finland',
 'Fiji',
 'France',
 'Micronesia',
 'Gabon',
 'United Kingdom',
 'Georgia',
 'Ghana',
 'Gambia',
 'Greece',
 'Honduras',
 'Croatia',
 'Haiti',
 'Hungary',
 'Indonesia',
 'India',
 'Ireland',
 'Iran',
 'Iraq',
 'Iceland',
 'Israel',
 'Italy',
 'Jamaica',
 'Jordan',
 'Japan',
 'Kazakhstan',
 'Kenya',
 'Kyrgyzstan',
 'Cambodia',
 'Kiribati',
 'Saint Kitts and Nevis',
 'South Korea',
 'Kuwait',
 'Laos',
 'Lebanon',
 'Liberia',
 'Liechtenstein',
 'Sri Lanka',
 'Lesotho